### DISCUSSION TITLE
MS Word weirdness

### DISCUSSION THREAD

**Discussion by DP-Upstate** [psb1558/Junicode-font#225] [General]
I recently discovered the new version of Junicode that is under development, and I am delighted!  It is a wonderful font and many thanks to Peter and everyone else who has contributed.

I use Windows 11.  The fonts I downloaded work perfectly (as far as I can tell so far) in LibreOffice Writer and Affinity Publisher.  In MS Word 2016 I get the following:

![Junicode screen shot](https://github.com/psb1558/Junicode-font/assets/53433504/c3d17167-95a0-40b7-9306-c5a86af7fd3c)

The last line has Junicode Beta applied.  This so bizarre.  I thought about filing this as a bug report (and will do so if that's helpful) but decided to ask about it more generally first.  It may be a Word issue, but still . . . The language of this document is US English, so there's nothing that should trigger odd behavior.

**Comment by psb1558** [comment]
Which version of the font are you using? Static TTF, Static OTF, or variable (I assume the version number is 2.000beta7)?

Did you have an earlier version of Junicode installed? If so, the earlier version may be hanging around in the font cache. [Here's](https://winaero.com/rebuild-font-cache-windows-10/) an article on rebuilding the font cache in Windows 10—but others here know a lot more about Windows than I do.

**Reply by DP-Upstate** [reply]
I'm using OTF static.  I had not used the earlier Junicode in a long time, even though it is still installed on my system, so it's unlikely to be a font cache issue.  But I will uninstall it and see what happens.  It still works (just tried it), and the two fonts are displayed with different names (Junicode vs Junicode Two Beta) -- so I assume that Windows sees them as different fonts.  And, as I said, the Beta works correctly in other apps.

**Reply by DP-Upstate** [reply]
Just uninstalled Junicode and rebooted.  Same result in Word with Junicode Two Beta.

**Comment by psb1558** [comment]
Issue #204 may be related to what you're seeing, though that user is not experiencing problems as extreme as yours. But click over there to see the result of my experiment when I tried out Junicode 2 in Word on my son's Windows 11 computer.

One observation: I changed the font's family name from "Junicode Two Beta" to "Junicode" seven versions ago, which means that the copy you're testing is quite old. I've corrected many problems since then. I'm now building version 2.000beta8, and I'll let you know when it's ready. Install that (making sure to uninstall any copies of Junicode or Junicode Two Beta) and see if it makes a difference.

**Comment by kenmcd** [comment]
This demo is from the v2.000beta7 TTF files - Word 2019 - Windows 10
Exported from Word to PDF, then exported from the PDF to PNG.

![Quick brown fox 2 000beta7 from Word 2019 on Win 10_1](https://github.com/psb1558/Junicode-font/assets/844920/048ce7b1-1e4d-4339-8c63-c60beda3bbec)

**Comment by psb1558** [comment]
But when I test with the OTF version in MS Word on my son's Windows 11 machine, I see exactly what @DP-Upstate is seeing. Delete the font cache and restart, and the result is the same.

Perform the same experiment in LibreOffice (in LO/Windows, I see, I have access to *all the styles*—refreshing!), and everything is fine. So something odd is going on between Junicode OTF and Word.

Or did I perhaps fail to clear the font cache? Could there be, maybe, a user-level font cache in addition to the system-level one?

Because what this looks like to me is some piece of programming (Word or some system service that Word is calling but LibreOffice is not) consulting the wrong cmap.

**Comment by kenmcd** [comment]
> So something odd is going on between Junicode OTF and Word.

OK. Will also fix the OTF SemiBolds and then test all the OTFs with Word the same way.
That is kinda strange that you both have this issue with the OTFs.

**Reply by psb1558** [reply]
You've now got beta8 with the stylename fix.

**Reply by kenmcd** [reply]
> You've now got beta8 with the stylename fix.

OK. Thanks. I will go get those for further testing.

**Comment by kenmcd** [comment]
Well this is interesting...
Uninstalled the TTFs, and installed the OTFs. v2.000beta7
Still Word 2019 on Windows 10. Same docx as above.

![Quick brown fox 2 000beta7 from Word 2019 on Win 10-OTF fonts](https://github.com/psb1558/Junicode-font/assets/844920/7a8917fa-24d9-406f-8987-36605066ba69)

**Comment by kenmcd** [comment]
The same DOCX file opened in LibreOffice with the same OTF fonts - looks fine.
Note: LibreOffice does build its own font cache.
I will try clearing my Win10 font cache, but I do not think that is the issue.
This is again an image of the PDF export.

![Quick brown fox 2 000beta7 from LibreOffice on Windows 10_1](https://github.com/psb1558/Junicode-font/assets/844920/124cc57f-0095-47f8-bd76-956ff464f94f)

**Comment by psb1558** [comment]
Yup. Exactly what I'm seeing. And if it goes for you as it did for me, clearing the cache won't help at all.

I'm going to crack open one of the OTFs and look at the cmaps.

**Reply by kenmcd** [reply]
> Yup. Exactly what I'm seeing. And if it goes for you as it did for me, clearing the cache won't help at all.
> 
> I'm going to crack open one of the OTFs and look at the cmaps.

Yes, that does not look like font cache corruption - too perfect, too uniform.
It is like everything is shifted the same way.
Hmmm... I am going to look at the codes behind those `.notdef` characters in the PDF.

**Comment by psb1558** [comment]
There are four cmap subtables: two (formats 4 and 12) for platform 0 ("Unicode") and two (same formats) for platform 3 ("Windows"). At a glance, they all look okay. Not scrambled or anything.

Format 12 is "the standard character-to-glyph-index mapping subtable for fonts supporting Unicode character repertoires that include supplementary-plane characters (U+10000 to U+10FFFF)." Junicode uses the supplementary plane because of Gothic and some other stuff, but it's hard to imagine that it could be a problem: after all, the emojis are encoded there, and there would be wailing and gnashing to teeth if Word messed *those* up.

I'm not finding any relevant issues on the fontmake site.

**Comment by kenmcd** [comment]
Very strange PDF output from Word.
Only two fonts embedded - Brill-Italic and Brill-Roman. 
Both embedded as TrueType fonts.
Went back and looked at the DOCX, and all the correct Junicode fonts are assigned.
The DOCX has not been changed since the TTFs were output correctly.

Hmmm... just remembered - Word does not embed OTF fonts in PDFs, or the DOCX files... but I do not remember it not even displaying the fonts. 
IIRC it does print properly.
Maybe what I am remembering is just related to embedding OTF fonts in DOCX files. 
Have to go find this info about Word and OTF fonts.

**Comment by psb1558** [comment]
I wonder if I should just drop the `.otf` fonts. IIRC, I started generating them because some printer drivers were not displaying composite glyphs correctly. I didn't understand what the issue was at that time. Now I understand, and have corrected.

And it seems that Word has had issues with `.otf` fonts in the relatively recent past, see e.g. [here](https://answers.microsoft.com/en-us/msoffice/forum/all/word-365-does-not-display-open-type-cff-fonts/5486c388-9807-475d-9fb8-de25acb29d05?page=2). I've always been a TrueType kind of guy; so what's the point of generating the font in a different format, adding to the hassle?

**Comment by kenmcd** [comment]
> And it seems that Word has had issues with `.otf` fonts in the relatively recent past, see e.g. [here](https://answers.microsoft.com/en-us/msoffice/forum/all/word-365-does-not-display-open-type-cff-fonts/5486c388-9807-475d-9fb8-de25acb29d05?page=2). I've always been a TrueType kind of guy; so what's the point of generating the font in a different format, adding to the hassle?

Yes, I found another example of Word 365 not displaying OTF fonts.
https://techcommunity.microsoft.com/t5/word/word-365-does-not-display-otf-fonts/m-p/2797350
But those seem to apply to all OTF fonts.
I checked a few other OTF fonts that I already have installed, and they work fine.
Akzidenz Grotesk Next, Helvetica Now, Garamond Premier Pro, Frutiger Next LT W1G, Nimbus Roman, Proxima Nova, Raleway, Roboto Flex OTF statics (have no idea why I have these installed).
Anyway you get the idea. They all seem to work in Word.

I did just notice in the Windows Font Manager that the Junicode OTF fonts are shown in seven different families - and these families do not make any sense.
So Word could just be confused.
Obviously the Windows Font Manager is confused.

Earlier today I did check the name fields at the top of the `CFF` table in the OTF fonts as those name fields are sometimes used by applications - so if they are missing or wrong they can cause issues.
This recently happened with LibreOffice and a couple of the Source font families.
But the Junicode fonts I looked at seemed to be OK. 
Nothing was missing and the names were the same as in the `name` table.

I did make some reorganized TTF fonts to test in different applications.
Made separate typographic families for widths, so they matched the style groups.
Have not tested them yet. 
Renamed it Junitest so they can be installed at the same time for comparison.
Maybe I should do the same for the OTF fonts.
See if they work in Word, and if the Windows Font Manager looks sane.

Edit: and I did confirm Office apps will not embed OTF fonts in Office docs.

**Comment by DP-Upstate** [comment]
Thanks for doing all this checking, guys.  This is a real pain, but I do feel relieved that I wasn't doing something stupid to cause the problem on my end.  I have version 1.064 OTF installed.  I only found Junicode Two about three weeks ago, and I thought I downloaded the most recent (at the time).  I will now go and grab the TTFs and see if they work correctly in Word.

**Comment by DP-Upstate** [comment]
OK, uninstalled the 1.064 OTF and put in 2.000-8 TTF.  These work in Word, which is good.  Two things:

a. I think the OTF fonts looked better on screen, while the TTFs look a little thin/washed out.  I can't go back and compare directly now, unfortunately.
b. Windows put the Semibold fonts in a separate family folder.  With the previous OTFs and also with other fonts I have such as Minion Pro and Libertinus Serif, the semibold is in the main family folder.  I don't know if this matters, as long as the fonts work.  (I understand of course that Word and many other Windows apps only understand four-weight families and so one has to select semibold as a separate font.)
Update: I had to go back into Affinity Publisher and update some styles which had Semibold earlier and now need SmBold.  Based on other fonts I have, Semibold seems to be the more common name for this variant.

**Reply by psb1558** [reply]
Yes, TrueType and CFF fonts will look different on screen. The outlines should be the same (functionally, anyway—the two kinds of font use different kinds of outlines), but the hinting systems are radically different. For the static fonts I use a program called ttfautohint to hint the TrueType fonts. It's not perfect, but it is the only game in town, and I can't manually hint 38 font files.

The variable fonts, on the other hand, are manually (lovingly) hinted. But they don't work in Word yet, right?

Best, I guess, to put up with slightly spindly letters than gibberish on screen.

**Reply by psb1558** [reply]
I'll have a look at how Libertinus organizes its font faces. If their system seems better, I'll adopt it.

**Reply by DP-Upstate** [reply]
Yes, now that you mention it, I remember the different hinting systems between OTF and TTF fonts.  Of course spindly letters are better than gibberish.

This is what I see in the Windows font folder for Libertinus Serif, if it helps:
![image](https://github.com/psb1558/Junicode-font/assets/53433504/ce891707-7fba-4780-b1d9-825691c4bf56)

**Comment by kenmcd** [comment]
@psb1558 
Think I have a solution for the OTF fonts in Word.

Short version:
Remove the `Symbol Character Set` from the Code Pages and it works in Word.

Long version:
FontLab would not export the OpenType features.
It would show some error messages, and then basically say it was giving-up.
It would still export the font, but no OpenType features at all, not even kerning.
So I exported the new v2.000 Junicode-Regular OTF to test and see if it worked in Word.
To see if the issues were in the OpenType features (which are gone in this font).
Installed this stripped font.
Nope. Still garbage in Word.

Then I opened the v2.000 Junicode-Regular OTF in FontCreator and used it to temporarily install the font - and it worked in Word.
So I exported the font and then actually installed that font.
That font sorta worked - it displayed various symbols instead of the text.
Something is going on with the symbols encoding.
I had noticed the `Symbol Character Set` in the Code Pages the other day.
So I deleted the `Symbol Character Set` in the Code Pages, exported the font again,
and it worked in Word.

Went back to my  v2.000 Junicode-Regular OTF font with no OpenType code in FontLab,
deleted the `Symbol Character Set` in the Code Pages, exported the font again,
and it now also worked in Word.

So I think Word is being confused by the `Symbol Character Set` in the Code Pages.

**Comment by psb1558** [comment]
Good God, how extremely strange! I'll try generating the font without that code page and see what happens.

**Comment by DP-Upstate** [comment]
This actually makes sense to me, sort of. I remember from like 30 years ago when fonts were limited to 256 characters. The Symbol code page had the usual letters and numbers replaced with a specific set of symbol glyphs defined by IBM. I remember reading in the manual for some font software (I think) that one should never check the symbol code page unless making that specific kind of font.On Aug 17, 2023, at 9:59 PM, Peter Baker ***@***.***> wrote:﻿
Good God, how extremely strange! I'll try generating the font without that code page and see what happens.

**Comment by psb1558** [comment]
Well, there is it:
<img width="686" alt="image" src="https://github.com/psb1558/Junicode-font/assets/16215767/f330cdcf-d2db-4998-b3f6-c33049e1037c">
What an amazing piece of detective work, @kenmcd!

And what an astonishing thing that one bit in this ancient structure can make such a difference in a supposedly modern app like MS Word. No wonder they can't manage to implement small caps or a decent collection of OpenType features—ancient font technology must permeate the code base.

**Comment by kenmcd** [comment]
> I remember reading in the manual for some font software (I think) that one should never check the symbol code page unless making that specific kind of font.

@DP-Upstate 
I checked the manuals for FontCreator, FontLab, and Glyphs - no luck.
Do you happen to use some other font editor where you may have seen that?

@psb1558 
I did find some related info in the FontLab manual about OpenType Symbol Fonts which does point out some of the oddities with Windows "special handling", different CMAP formats, differences in handling OTF fonts vs. TTF fonts, etc.
It does not specifically discuss this exact thing, but it is obviously related to the problem.
Thought you may want to take a look.
The older v6 manual is easier to read online (the FL7/8 manual format sucks).
https://help.fontlab.com/fontlab-vi/Release-Notes-Archive/#opentype-symbol-fonts

**Comment by psb1558** [comment]
Thanks, @kenmcd, I'll have a look.

Meanwhile, that one bit makes such a difference that I've made a new release because of it. This one includes the `.otf` files.
